File: solutions/scala/165-path-sum-iii.scala
/**
 * Problem 165: Path Sum III (LeetCode 437)
 * Difficulty: Med
 * Language: Scala
 */
case class TreeNode(value: Int, left: Option[TreeNode] = None, right: Option[TreeNode] = None)

def pathSum(root: Option[TreeNode], target: Int): Int = {
  def dfs(node: Option[TreeNode], prefix: Map[Long, Int], curr: Long): Int = node match {
    case None => 0
    case Some(n) =>
      val sum = curr + n.value
      val count = prefix.getOrElse(sum - target, 0)
      val prefix2 = prefix.updated(sum, prefix.getOrElse(sum, 0) + 1)
      count + dfs(n.left, prefix2, sum) + dfs(n.right, prefix2, sum)
  }
  dfs(root, Map(0L -> 1), 0L)
}

File: solutions/rust/165-path-sum-iii.rs
/**
 * Problem 165: Path Sum III (LeetCode 437)
 * Difficulty: Med
 * Language: Rust
 */
use std::collections::HashMap;
use std::rc::Rc;
use std::cell::RefCell;

type TreeLink = Option<Rc<RefCell<TreeNode>>>;
struct TreeNode { val: i32, left: TreeLink, right: TreeLink }

fn path_sum(root: &TreeLink, target: i64) -> i32 {
    fn dfs(node: &TreeLink, prefix: &mut HashMap<i64, i32>, curr: i64, target: i64) -> i32 {
        match node {
            None => 0,
            Some(n) => {
                let n = n.borrow();
                let sum = curr + n.val as i64;
                let count = *prefix.get(&(sum - target)).unwrap_or(&0);
                *prefix.entry(sum).or_insert(0) += 1;
                let res = count + dfs(&n.left, prefix, sum, target)
                                + dfs(&n.right, prefix, sum, target);
                *prefix.get_mut(&sum).unwrap() -= 1;
                res
            }
        }
    }
    let mut prefix = HashMap::from([(0i64, 1)]);
    dfs(root, &mut prefix, 0, target)
}
